### ISSUE TITLE
Issues with union types in subscriptions

### ISSUE THREAD

**Issue by hognevevle** [ChilliCream/hotchocolate#1598]
Per https://hotchocolategraphql.slack.com/archives/CD9TNKT8T/p1584999380445000

Trying to use a union type in a subscription, but seeing some weird behaviour when I'm pushing some data to the topic through a mutation.

In BCP, I see nothing happening.

In playground, I see the following message being received: {id: "1", type: "complete"} , and the subscription is stopped.

![image](https://user-images.githubusercontent.com/2721188/77368416-75dc1d80-6d5c-11ea-99d8-e795efeca864.png)

IConversationChange.cs
```
    [UnionType(Name = "ConversationChange")]
    public interface IConversationChange
    {
    }
```

ConversationViewer.cs
```
    public class ConversationViewer : IConversationChange
    {
        public User User { get; set; }
        public DateTime LastSeenAt { get; set; }

        public ConversationViewer(User user)
        {
            User = user;
            LastSeenAt = DateTime.UtcNow;
        }
    } 
```

ConversationSubscriptions.cs
```
    [ExtendObjectType(Name = "Subscription")]
    public class ConversationSubscriptions
    {
        [SubscribeAndResolve]
        public async Task<IAsyncEnumerable<IConversationChange>> ConversationUpdatesAsync(
            int conversationId,
            [Service]ITopicEventReceiver eventReceiver,
            CancellationToken cancellationToken) {
            return await eventReceiver.SubscribeAsync<string, IConversationChange>(
                    $"conversation-changes_{conversationId}", cancellationToken)
                .ConfigureAwait(false);
        }
    }
```

ConversationMutations.cs
```
    public class ConversationMutations
    { 
        public async Task<ConversationViewPayload> ConversationViewAsync(
            ConversationViewInput input,
            [GlobalState]string currentUserToken,
            UserByTokenDataLoader userByTokenLoader,
            [Service]ITopicEventSender eventSender,
            CancellationToken cancellationToken)
        {
            var users = await userByTokenLoader.LoadAsync(cancellationToken, currentUserToken);

            if (!users?.Any() ?? true)
            {
                throw new Exception("User not found");
            }

            var msg = new ConversationViewer(users.First());

            await eventSender.SendAsync(
                $"conversation-changes_{input.ConversationId}", msg, cancellationToken);

            return new ConversationViewPayload();
        }
    }
```

Mutation:
```
mutation {
  conversationView(input: {conversationId: 1055}) { __typename }
}
```

Subscription:
```
subscription {
  conversationUpdates(conversationId:1055) {
        ...on ConversationViewer {
          lastSeenAt
          user { name }
        }
      }
}
```

**Comment by hognevevle** [comment]
I originally tried this only with Redis, but I've reproduced it with in-memory subscriptions too.

For in-memory, it worked after casting `msg` to `IConversationChange` in SendAsync.

When doing the same with Redis, however, the message isn't properly being sent through. (redis monitor shows '{}' being sent). 

The only way I've been able to make subscriptions work with Redis so far, is by using simple string messages. It might look like a deserialization issue in the Redis implementation when using object messages. 

Is there any point in fixing this for v10, or will all of this be rewritten in v11?

**Comment by hognevevle** [comment]
@michaelstaib Repro at https://github.com/hognevevle/HC_1598_repro

**Comment by hognevevle** [comment]
I suspect the JSON serialization happening when the message is being sent through Redis to be the underlying issue here.

I was able to work around it by converting the message payload into base64 (then simply sending/receiving it as a `string`). Then, on the receiving end, deserialize it back to an object. This way, it works fine. 

Not sure if `System.Text.Json` supports the full blown deserialization/serialization which we need here (I know Newtonsoft allows you to serialize in a way which also embeds type details in the serialized data)..

**Comment by michaelstaib** [comment]
Yes it is and it is fixed ... in my 11 execution engine branch .... I will port the fix back to 10.5

**Comment by michaelstaib** [comment]
This one is fixed with 10.5.rc.0
